pkg/ensurance/node-resource/node-resource-manager.go

```diff
@@ -0,0 +1,366 @@
+				continue
+			}
+			node.Status.Capacity[v1.ResourceName(extResourceName)] =
+				*resource.NewQuantity(int64(nextRecommendation), resource.DecimalSI)
+			node.Status.Allocatable[v1.ResourceName(extResourceName)] =
+				*resource.NewQuantity(int64(nextRecommendation), resource.DecimalSI)
+		}
+	}
+}
+
+func (o *NodeResourceManager) reconcilePod(obj interface{}) {
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("Couldn't get object from tombstone %#v", obj))
+			return
+		}
+		pod, ok = tombstone.Obj.(*v1.Pod)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("Tombstone contained object that is not a pod %#v", obj))
+			return
+		}
+	}
+
+	o.updatePodExtResToCgroup(pod)
+}
+
+func ownedPod(pod *v1.Pod, nodeName string) bool {
+	return pod.Spec.NodeName == nodeName && pod.Status.Phase != v1.PodSucceeded && pod.Status.Phase != v1.PodFailed
+}
+
+// Get pod's ext-resource.node.gocrane.io resource and update it to Cgroup
+func (o *NodeResourceManager) updatePodExtResToCgroup(pod *v1.Pod) {
+	for _, c := range pod.Spec.Containers {
+		for res, val := range c.Resources.Limits {
+			if strings.HasPrefix(res.String(), fmt.Sprintf(ExtResourcePrefixFormat, v1.ResourceCPU)) {
+				containerId := getContainerIdFromPod(pod, c.Name)
+				if containerId == "" {
+					continue
+				}
+
+				containerPeriod := o.getCPUPeriod(pod, containerId)
+				if containerPeriod == 0 {
+					continue
+				}
+
+				// Update cpu quota by CRI
+				err := cruntime.UpdateContainerResources(o.runtimeClient, containerId, cruntime.UpdateOptions{CPUQuota: int64(float64(val.MilliValue()) / executor.CpuQuotaCoefficient * containerPeriod)})
+				if err != nil {
+					klog.Errorf("Failed to update pod %s container %s Resource, err %s", pod.Name, containerId, err.Error())
+					continue
+				}
+			}
+		}
+	}
+}
+
+func getContainerIdFromPod(pod *v1.Pod, containerName string) string {
+	for _, cs := range pod.Status.ContainerStatuses {
+		if cs.Name == containerName {
+			return utils.GetContainerIdFromKey(cs.ContainerID)
+		}
+	}
+	return ""
+}
+
+// Get cpu period from local if cadvisor collector exists and state is not expired;
+// Otherwise, get value from CRI
+func (o *NodeResourceManager) getCPUPeriod(pod *v1.Pod, containerId string) float64 {
+	value, exists := o.collectors.Load(types.CadvisorCollectorType)
+	now := time.Now()
+
+	if exists && o.state != nil && !now.After(o.lastStateTime.Add(StateExpiration)) {
+		_, containerCPUPeriods := executor.GetPodUsage(string(stypes.MetricNameContainerCpuPeriod), o.state, pod)
+		for _, period := range containerCPUPeriods {
+			if period.ContainerId == containerId {
+				return period.Value
+			}
+		}
+	}
+
+	// Use CRI to get cpu period directly
+	c := value.(cadvisor.CadvisorCollector)
+	var query = info.ContainerInfoRequest{}
+	containerInfoV1, err := c.Manager.GetContainerInfo(containerId, &query)
```

Comment:
Not get state direct, we can use the state

Reply:
We inded get cpu period from local if cadvisor collector exists and state is not expired first.
If state is not available, get it from CRI.